src/runtime-rs/crates/hypervisor/src/qemu/cmdline_generator.rs

```diff
@@ -1824,6 +1825,178 @@ impl ToQemuParams for ObjectSevSnpGuest {
     }
 }
 
+/// PCIeRootPortDevice directly attached onto the root bus
+/// -device pcie-root-port,id=rp0,bus=pcie.0,chassis=0,slot=0,multifunction=off,pref64-reserve=<X>B,mem-reserve=<Y>B
+#[derive(Debug, Default)]
+pub struct PCIeRootPortDevice {
+    id: String,
+    bus: String,
+    chassis: String,
+    slot: String,
+    multifunction: bool,
+    addr: String,
+
+    mem_reserve: String,
+    pref64_reserve: String,
+}
+
+impl PCIeRootPortDevice {
+    fn new(id: &str, bus: &str, chassis: &str, slot: &str, multifunc: bool, addr: &str) -> Self {
+        PCIeRootPortDevice {
+            id: id.to_string(),
+            bus: if bus.is_empty() {
+                DEFAULT_PCIE_ROOT_BUS.to_owned()
+            } else {
+                bus.to_string()
+            },
+            chassis: chassis.to_string(),
+            slot: slot.to_string(),
+            multifunction: multifunc,
+            addr: addr.to_string(),
+            ..Default::default()
+        }
+    }
+
+    fn set_mem_reserve(&mut self, mem_reserve: u64) -> &mut Self {
+        if mem_reserve > 0 {
+            self.mem_reserve = format!("{}B", mem_reserve);
+        }
+
+        self
+    }
+
+    fn set_pref64_reserve(&mut self, pref64_reserve: u64) -> &mut Self {
+        if pref64_reserve > 0 {
+            self.pref64_reserve = format!("{}B", pref64_reserve);
+        }
+
+        self
+    }
+}
+
+#[async_trait]
+impl ToQemuParams for PCIeRootPortDevice {
+    async fn qemu_params(&self) -> Result<Vec<String>> {
+        let mut device_params = Vec::new();
+
+        // -device pcie-root-port,id=rp0
+        device_params.push(format!("{},id={}", "pcie-root-port", self.id));
+
+        let bus = if self.bus.is_empty() {
+            DEFAULT_PCIE_ROOT_BUS
+        } else {
+            &self.bus
+        };
+        device_params.push(format!("bus={}", bus));
+
+        let chassis = if self.chassis.is_empty() {
+            "0x00"
+        } else {
+            &self.chassis
+        };
+        device_params.push(format!("chassis={}", chassis));
+
+        let slot = if self.slot.is_empty() {
+            "0x00"
+        } else {
+            &self.slot
+        };
+        device_params.push(format!("slot={}", slot));
+
+        let multifunction = if self.multifunction { "on" } else { "off" };
+        if self.multifunction {
+            let addr = if self.addr.is_empty() {
+                "0x00"
+            } else {
+                &self.addr
+            };
+            device_params.push(format!("addr={}", addr));
+        }
+
+        device_params.push(format!("multifunction={}", multifunction));
```

Comment:
Frankly I'd prefer for `chassis`, `slot`, `multifunction` and `addr` to be treated similar to `bus`, i.e. do the checking and default value substitution as early as possible, in `new()` since all of them are non-optional params of `new()`.

However I don't see this as blocking a merge, I can make these adjustments later on if we agree on doing so.

Reply:
> Frankly I'd prefer for `chassis`, `slot`, `multifunction` and `addr` to be treated similar to `bus`, i.e. do the checking and default value substitution as early as possible, in `new()` since all of them are non-optional params of `new()`.
> 
Good cache ! I think this is a very constructive and helpful suggestion. fixed

Reply:
Thanks!